0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md

```diff
@@ -1,4 +1,263 @@
+`C:\Users\j3qq4hch>arm-none-eabi-gdb`
+Then connect to DCP load your firmware and run it
+
+```
+(gdb) tar ext COM6
+Remote debugging using COM6
+(gdb) mon swdp_scan
+Target voltage: 4.4V
+Available Targets:
+No. Att Driver
+1  STM32F3
+(gdb) file C:\\zubax_babel\\firmware\\build\\compound.elf
+Reading symbols from C:\zubax_babel\firmware\build\compound.elf...warning: Loadable section "bootloader" outside of ELF segments
+done.
+(gdb) attach 1
+Attaching to program: C:\zubax_babel\firmware\build\compound.elf, Remote target
+0x0800d9fe in watchdogReset (id=id@entry=0) at zubax_chibios///zubax_chibios/platform/stm32/watchdog_stm32.cpp:139
+139 if ((_mask & valid_bits_mask) == valid_bits_mask)
+(gdb) load
+Loading section bootloader, size 0x5571 lma 0x8000000
+Loading section startup, size 0x1c0 lma 0x8008000
+Loading section constructors, size 0x4 lma 0x80081c0
+Loading section .padding1, size 0x1c lma 0x80081c4
+Loading section .text, size 0x7c48 lma 0x80081e0
+Loading section .data, size 0x90 lma 0x800fe28
+Loading section .noinit, size 0x8 lma 0x800feb8
+Start address 0x80081e0, load size 54321
+Transfer rate: 16 KB/sec, 890 bytes/write.
+(gdb) run
+The program being debugged has been started already.
+Start it from the beginning? (y or n) y
+Starting program: C:\zubax_babel\firmware\build\compound.elf
+```
+### Loading with bootloader ###
+coming soon
+
+----------
+## Loading firmware in *nix environment ##
+### Loading with DCP ###
+### Loading with bootloader ###
+----------
+
+## Tutorials ##
+
+### Blink (in C) ###
+
+A couple words about general code organisation must be said first.
+Zubax Babel firmware uses ChibiOS and therefore this tutorial also relies on ChibiOS and its HAL in particular.
+Firmware is written in C++, but it is OK to write in C too. 
+As in any OS code is organised as threads which are executed in a pseudo-parallel way. Each thread is represented as separate endless function, which is called by the OS-scheduler according to its rules and desires. If you are interested how ChibiOS works (and even if not), it is highly recommended to read tutorial and manuals from [here](http://www.chibios.org/dokuwiki/doku.php?id=chibios:documentation:start). 
+
+Finally, lets start.
+We should begin with writing HELLO_WORLD for MCU - blinking an onboard LED. According to babel schematic one of the LEDs is connected to `pin8` of `PORTE`. Let’s blink it. Create blink function. Something like this:
+
+
+```c
+static THD_WORKING_AREA(waThread1, 128);
+static THD_FUNCTION(blink_thread, arg) 
+{   
+    (void)arg;
+    chRegSetThreadName("blinker"); 
+    palSetPadMode(GPIOE, 8, PAL_MODE_OUTPUT_PUSHPULL);
+    while (true) 
+    {
+        palSetPad(GPIOE,8);
+        chThdSleepMilliseconds(100);
+        palClearPad(GPIOE, 8);
+        chThdSleepMilliseconds(100);
+    }
+}
+```
+Now find function ```int main()``` in ```zubax_babel\firmware\src\main.cpp```. This is program entry point. Comment it all and write your own version like this:
```

Comment:
Here and elsewhere: inline code blocks are highlighted with single backticks: ``` `code` ```. Triple backticks are needed to highlight multi-line code blocks.